netket/utils/group/group.py

```diff
@@ -0,0 +1,329 @@
+            representatives: the lowest-indexed member of each conjugacy class
+            inverse: the conjugacy class index of every group element
+        """
+        row_index = np.arange(len(self))[:, np.newaxis]
+
+        # if is_conjugate[i,j] is True, self[i] and self[j] are in the same class
+        is_conjugate = np.full((len(self), len(self)), False)
+        is_conjugate[row_index, self.conjugacy_table] = True
+
+        classes, representatives, inverse = np.unique(
+            is_conjugate, axis=0, return_index=True, return_inverse=True
+        )
+
+        # Usually self[0] == Identity(), which is its own class
+        # This class is listed last by the lexicographic order used by np.unique
+        # so we reverse it to get a more conventional layout
+        classes = classes[::-1]
+        representatives = representatives[::-1]
+        inverse = (representatives.size - 1) - inverse
+
+        return classes, representatives, inverse
+
+    @struct.property_cached
+    def character_table_by_class(self) -> Array:
+        r"""
+        Calculates the character table using Burnside's algorithm.
+
+        Each row of the output lists the characters of one irrep in the order the
+        conjugacy classes are listed in `self.conjugacy_classes`.
+
+        Assumes that `Identity() == self[0]`, if not, the sign of some characters
+        may be flipped. The irreps are sorted by dimension.
+        """
+        classes, _, _ = self.conjugacy_classes
+        class_sizes = classes.sum(axis=1)
+        # Construct a random linear combination of the class matrices c_S
+        #    (c_S)_{RT} = #{r,s: r \in R, s \in S: rs = t}
+        # for conjugacy classes R,S,T, and a fixed t \in T.
+        #
+        # From our oblique times table it is easier to calculate
+        #    (d_S)_{RT} = #{r,t: r \in R, t \in T: rs = t}
+        # for a fixed s \in S. This is just `product_table == s`, aggregrated
+        # over conjugacy classes. c_S and d_S are related by
+        #    c_{RST} = |S| d_{RST} / |T|;
+        # since we only want a random linear combination, we forget about the
+        # constant |S| and only divide each column through with the appropriate |T|
+        class_matrix = (
+            classes @ np.random.uniform(size=len(self))[self.product_table] @ classes.T
+        )
+        class_matrix /= class_sizes
+
+        # The vectors |R|\chi(r) are (column) eigenvectors of all class matrices
+        # the random linear combination ensures (with probability 1) that
+        # none of them are degenerate
+        _, table = np.linalg.eig(class_matrix)
+        table = table.T / class_sizes
+
+        # Normalise the eigenvectors by orthogonality: \sum_g |\chi(g)|^2 = |G|
+        norm = np.sum(np.abs(table) ** 2 * class_sizes, axis=1, keepdims=True) ** 0.5
+        table /= norm
+        table /= _cplx_sign(table[:, 0])[:, np.newaxis]  # ensure correct sign
+        table *= len(self) ** 0.5
```

Comment:
I may be completely wrong, but wrt our numerical stability issues in the tests I am a bit suspicious about this normalization code here. Maybe there is some cancellation occurring here when dividing out the norm... is there an alternative way of doing this normalization we could try for comparison?

Reply:
The code above looks fine to me.
Probably it's not the most accurate.
I just found out that `np.sum` does not use floating-point stabilization tricks to increase precision, so it's somewhat... imprecise, especially for large matrices.

However, I doubt we'd need that kind of precision other than in those tests...